pkg/vm/stackitem/json.go

```diff
@@ -260,72 +261,120 @@ func (d *decoder) decodeMap() (*Map, error) {
 
 // ToJSONWithTypes serializes any stackitem to JSON in a lossless way.
 func ToJSONWithTypes(item Item) ([]byte, error) {
-	result, err := toJSONWithTypes(item, make(map[Item]bool, typicalNumOfItems))
-	if err != nil {
-		return nil, err
-	}
-	return json.Marshal(result)
+	return toJSONWithTypes(nil, item, make(map[Item]sliceNoPointer, typicalNumOfItems))
 }
 
-func toJSONWithTypes(item Item, seen map[Item]bool) (interface{}, error) {
-	if len(seen) > MaxJSONDepth {
-		return "", ErrTooDeep
+func toJSONWithTypes(data []byte, item Item, seen map[Item]sliceNoPointer) ([]byte, error) {
+	if item == nil {
+		return nil, fmt.Errorf("%w: nil", ErrUnserializable)
+	}
+	if old, ok := seen[item]; ok {
+		if old.end == 0 {
+			// Compound item marshaling which has not yet finished.
+			return nil, ErrRecursive
+		}
+		if len(data)+old.end-old.start > MaxSize {
+			return nil, errTooBigSize
+		}
+		return append(data, data[old.start:old.end]...), nil
 	}
-	var value interface{}
+
+	switch item.(type) {
+	case Null:
+		if len(data)+14 > MaxSize {
+			return nil, errTooBigSize
+		}
+		data = append(data, `{"type":"Any"}`...)
+		return data, nil
+	case *Interop:
+		if len(data)+18 > MaxSize {
+			return nil, errTooBigSize
+		}
+		data = append(data, `{"type":"Interop"}`...)
+		return data, nil
+	}
+
+	start := len(data)
+
+	data = append(data, `{"type":"`+item.Type().String()+`","value":`...)
+	if len(data) > MaxSize {
+		return nil, errTooBigSize
+	}
+
+	// Primitive stack items are appended after the switch
+	// to reduce the amount of size checks.
+	var primitive string
+	var isBuffer bool
+	var err error
+
 	switch it := item.(type) {
 	case *Array, *Struct:
-		if seen[item] {
-			return "", ErrRecursive
-		}
-		seen[item] = true
-		arr := []interface{}{}
-		for _, elem := range it.Value().([]Item) {
-			s, err := toJSONWithTypes(elem, seen)
+		seen[item] = sliceNoPointer{}
+		data = append(data, '[')
+		for i, elem := range it.Value().([]Item) {
+			if i != 0 {
+				data = append(data, ',')
+			}
+			data, err = toJSONWithTypes(data, elem, seen)
 			if err != nil {
-				return "", err
+				return nil, err
 			}
-			arr = append(arr, s)
 		}
-		value = arr
-		delete(seen, item)
 	case Bool:
-		value = bool(it)
-	case *Buffer, *ByteArray:
-		value = base64.StdEncoding.EncodeToString(it.Value().([]byte))
+		if it {
+			primitive = "true"
+		} else {
+			primitive = "false"
+		}
+	case *ByteArray:
+		primitive = `"` + base64.StdEncoding.EncodeToString(it.Value().([]byte)) + `"`
+	case *Buffer:
+		isBuffer = true
+		seen[item] = sliceNoPointer{}
```

Comment:
Not really needed (it won't be recursive and it's sent after the switch anyway)?

Reply:
Fixed